### PR TITLE
[Mu4] Midi output port

### PR DESCRIPTION
**Pull request by igorkorsukov**

### CODE REVIEW

File: mu4/appshell/qml/DevTools/DevToolsPage.qml [hunk 1]
```diff
@@ -73,7 +75,11 @@ DockPage {
 
     Component {
         id: synthSettingsComp
-
         SynthSettings {}
     }
+
+    Component {
+        id: midiPortsComp
+        MIdiPorts {}
```
Comment (RomanPudashkin) [inline mu4/appshell/qml/DevTools/DevToolsPage.qml:83]: **Mi**diPorts?

File: framework/midi/imidioutport.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    struct Device {
+        std::string id;
+        std::string name;
+    };
+
+    virtual std::vector<Device> devices() const = 0;
+
+    virtual Ret connect(const std::string& deviceID) = 0;
+    virtual void disconnect() = 0;
```
Comment (RomanPudashkin) [inline framework/midi/imidioutport.h:46]: Why disconnect and isConnected have no deviceId argument?

File: framework/midi/imidioutport.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    virtual ~IMidiOutPort() = default;
+
+    struct Device {
+        std::string id;
+        std::string name;
+    };
+
+    virtual std::vector<Device> devices() const = 0;
+
+    virtual Ret connect(const std::string& deviceID) = 0;
```
Comment (RomanPudashkin) [inline framework/midi/imidioutport.h:45]: Can we connect to multiple devices? What will happen if someone call connect several times with different deviceID?

File: framework/midi/internal/midiportdatasender.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        if (pos->first >= toTick) {
+            break;
+        }
+
+        const Event& event = pos->second;
+
+        if (SKIP_EVENTS.find(event.type) != SKIP_EVENTS.end()) {
+            // noop
```
Comment (RomanPudashkin) [inline framework/midi/internal/midiportdatasender.cpp:56]: Why not if (SKIP_EVENTS.find(event.type) == SKIP_EVENTS.end()) ?

File: framework/audio/internal/midisource.h [hunk 5]
```diff
@@ -61,6 +62,7 @@ class MidiSource : public IAudioSource
     std::string m_name;
     std::shared_ptr<SL> m_sl;
     std::shared_ptr<midi::ISequencer> m_seq;
+    midi::ISequencer::Context* m_seqContext = nullptr;
```
Comment (psmokotnin) [inline framework/audio/internal/midisource.h:65]: why not unique_ptr?

File: mu4/appshell/appshell.qrc [hunk 6]
```diff
@@ -19,5 +19,6 @@
         <file>qml/HomePage/AddonsContent.qml</file>
         <file>qml/DevTools/Audio/SynthSettings.qml</file>
         <file>qml/DevTools/Audio/SoundFontsPanel.qml</file>
+        <file>qml/DevTools/Audio/MIdiPorts.qml</file>
```
Comment (psmokotnin) [inline mu4/appshell/appshell.qrc:22]: the filename has mistake M**i**diPorts.qml

File: framework/midi/imidioutport.h (hunk 3 above)
Comment (igorkorsukov) [reply framework/midi/imidioutport.h:45]: > Can we connect to multiple devices? 
No

> What will happen if someone call connect several times with different deviceID?
There will be a disconnect and a connection to a new device

File: framework/midi/internal/midiportdatasender.cpp (hunk 4 above)
Comment (igorkorsukov) [reply framework/midi/internal/midiportdatasender.cpp:56]: It seems to me that this is a little more clear, if the event that we skip, then we do not do anything.

File: framework/audio/internal/midisource.h (hunk 5 above)
Comment (igorkorsukov) [reply framework/audio/internal/midisource.h]: Changed on shared_ptr (it is shared between class)

File: mu4/appshell/appshell.qrc (hunk 6 above)
Comment (igorkorsukov) [reply mu4/appshell/appshell.qrc]: fixed

File: mu4/appshell/qml/DevTools/DevToolsPage.qml (hunk 1 above)
Comment (igorkorsukov) [reply mu4/appshell/qml/DevTools/DevToolsPage.qml]: fixed

**Review by RomanPudashkin — APPROVED** [review]

**Review by psmokotnin — APPROVED** [review]

**Pull request merged**